filebeat/input/filestream/internal/input-logfile/manager.go

```diff
@@ -166,34 +167,53 @@ func (cim *InputManager) Create(config *common.Config) (input.Input, error) {
 		return nil, errNoInputRunner
 	}
 
+	sourceIdentifier := newSourceIdentifier(cim.Type, settings.ID)
+	pStore := cim.getRetainedStore()
+	defer pStore.Release()
+	prospectorStore := newSourceStore(pStore, sourceIdentifier)
+	err = prospector.Init(prospectorStore)
+	if err != nil {
+		return nil, err
+	}
+
 	return &managedInput{
-		manager:      cim,
-		prospector:   prospector,
-		harvester:    harvester,
-		cleanTimeout: settings.CleanTimeout,
+		manager:          cim,
+		userID:           settings.ID,
+		prospector:       prospector,
+		harvester:        harvester,
+		sourceIdentifier: sourceIdentifier,
+		cleanTimeout:     settings.CleanTimeout,
 	}, nil
 }
 
-// Lock locks a key for exclusive access and returns an resource that can be used to modify
-// the cursor state and unlock the key.
-func (cim *InputManager) lock(ctx input.Context, key string) (*resource, error) {
-	resource := cim.store.Get(key)
-	err := lockResource(ctx.Logger, resource, ctx.Cancelation)
-	if err != nil {
-		resource.Release()
-		return nil, err
-	}
-	return resource, nil
+func (cim *InputManager) getRetainedStore() *store {
+	store := cim.store
+	store.Retain()
+	return store
 }
 
-func lockResource(log *logp.Logger, resource *resource, canceler input.Canceler) error {
-	if !resource.lock.TryLock() {
-		log.Infof("Resource '%v' currently in use, waiting...", resource.key)
-		err := resource.lock.LockContext(canceler)
-		if err != nil {
-			log.Infof("Input for resource '%v' has been stopped while waiting", resource.key)
-			return err
-		}
+type sourceIdentifier struct {
+	prefix           string
+	configuredUserID bool
+}
+
+func newSourceIdentifier(pluginName, userID string) *sourceIdentifier {
+	inputPrefix := pluginName
+	configuredUserID := false
+	if userID != "" {
+		inputPrefix = inputPrefix + "::" + userID
+		configuredUserID = true
 	}
-	return nil
+	return &sourceIdentifier{
+		prefix:           inputPrefix + "::",
+		configuredUserID: configuredUserID,
```

Comment:
The check for `configuredUserID` is only used in `CleanIf`, but not in `UpdateIdentifier`. Plus `inputPrefix` is again a prefix of `inputPrefix::<id>`. That is `MatchesInput` will again return true, also it should return false.

How about breaking ties by introducing a `.global` id? E.g. all inputs instances that have no ID configured will use `<pluginName>::.global::`. If someone tries to set `.global` as ID in the configuration, we fail to create the input.

e.g.

```


func newSourceIdentifier(pluginName, inputID string) (*sourceIdentifier, error) {
  if inputID == ".global" { return nil, errors.New(...) }
  if inputID == "" { userID = ".global" }
  return &sourceIdentifier{ prefix: pluginName + "::" + inputID + "::" }, nil
}
```